File: DynamicAudioNormalizerPAS/src/DynamicAudioNormalizerExample.pas
//////////////////////////////////////////////////////////////////////////////////
// Dynamic Audio Normalizer - Delphi/Pascal Wrapper
// Copyright (c) 2014-2017 LoRd_MuldeR <mulder2@gmx.de>. Some rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//
// http://opensource.org/licenses/MIT
//////////////////////////////////////////////////////////////////////////////////

unit DynamicAudioNormalizerExample;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, AppEvnts, Math, DateUtils, ComCtrls, XPMan;

type
  TDynamicAudioNormalizerTestApp = class(TForm)
    ButtonTest1: TButton;
    ButtonExit: TButton;
    ButtonTest2: TButton;
    ApplicationEvents1: TApplicationEvents;
    ProgressBar1: TProgressBar;
    XPManifest1: TXPManifest;
    OpenDialog: TOpenDialog;
    SaveDialog: TSaveDialog;
    procedure ButtonTest1Click(Sender: TObject);
    procedure ButtonExitClick(Sender: TObject);
    procedure ButtonTest2Click(Sender: TObject);
    procedure ApplicationEvents1Exception(Sender: TObject; E: Exception);
    procedure FormCreate(Sender: TObject);
  private
    LastUpdate: TDateTime;
    procedure StartStopTest(const start: boolean);
    procedure UpdateProgress;
  public
    { Public-Deklarationen }
  end;

var
  DynamicAudioNormalizerTestApp: TDynamicAudioNormalizerTestApp;

implementation

uses
  DynamicAudioNormalizer;

{$R *.dfm}

//-----------------------------------------------------------------------------
// Log Callback
//-----------------------------------------------------------------------------

{WARNING: This function must use 'cdecl' calling convention!}

procedure loggingHandler(const level: LongInt; const text: PAnsiChar); cdecl;
var
  tag: String;
begin
  case level of
    0: tag := 'DBG';
    1: tag := 'WRN';
    2: tag := 'ERR';
  else
    tag := '???';
  end;

  WriteLn(Format('[DynAudNorm][%s] %s', [tag, text]));
end;

//-----------------------------------------------------------------------------
// Helper Functions
//-----------------------------------------------------------------------------

procedure TDynamicAudioNormalizerTestApp.UpdateProgress;
begin
  if ProgressBar1.Position < ProgressBar1.Max then
  begin
    ProgressBar1.StepBy(1);
    LastUpdate := Now;
  end else
  begin
    if MilliSecondsBetween(Now, LastUpdate) >= 997 then
    begin
      ProgressBar1.Position := 0;
      LastUpdate := Now;
    end
  end;
  Application.ProcessMessages;
end;

procedure TDynamicAudioNormalizerTestApp.StartStopTest(const start: boolean);
var
  i: Integer;
begin
  if start then
  begin
    for i := 0 to Self.ControlCount-1 do
    begin
      Self.Controls[i].Enabled := False;
    end;
    ProgressBar1.Position := 0;
  end else
  begin
    MessageBeep(1000);
    for i := 0 to Self.ControlCount-1 do
    begin
      Self.Controls[i].Enabled := True;
    end;
    ProgressBar1.Position := ProgressBar1.Max;
  end;
end;

//-----------------------------------------------------------------------------
// Misc Functions
//-----------------------------------------------------------------------------

procedure TDynamicAudioNormalizerTestApp.FormCreate(Sender: TObject);
var
  major, minor, patch: LongWord;
  date, time, compiler, arch: PAnsiChar;
  debug: LongBool;
begin
  LastUpdate := Now;

  TDynamicAudioNormalizer.GetVersionInfo(major, minor, patch);
  WriteLn(Format('Library Version: %s', [StringReplace(Format('%u.%02u-%u', [major, minor, patch]), ' ', '0', [rfReplaceAll])]));

  TDynamicAudioNormalizer.GetBuildInfo(date, time, compiler, arch, debug);
  WriteLn(Format('Build: Date=%s; Time=%s; Compiler=%s; Arch=%s; Debug=%s'#10, [date, time, compiler, arch, BoolToStr(debug)]));

  TDynamicAudioNormalizer.SetLogFunction(@loggingHandler);
end;

procedure TDynamicAudioNormalizerTestApp.ButtonExitClick(Sender: TObject);
begin
  Application.Terminate;
end;

procedure TDynamicAudioNormalizerTestApp.ApplicationEvents1Exception(Sender: TObject; E: Exception);
begin
  MessageBox(self.Handle, @E.Message[1], 'ERROR', MB_ICONERROR or MB_TOPMOST);
  ExitProcess(DWORD(-1));
end;

//-----------------------------------------------------------------------------
// TEST #1
//-----------------------------------------------------------------------------

procedure TDynamicAudioNormalizerTestApp.ButtonTest1Click(Sender: TObject);
const
  rounds = 97;
  instances = 16;
var
  normalizer: Array [1..instances] of TDynamicAudioNormalizer;
  i,k: Cardinal;
begin
  StartStopTest(True);

  for k := 1 to rounds do
  begin
    WriteLn('Round ' + IntToStr(k) + ' of ' + IntToStr(rounds) + '.');
    for i := 1 to instances do
    begin
      normalizer[i] := TDynamicAudioNormalizer.Create(2, 44100, 500, 31, 0.95, 10.0, 0.0, 0.0, True, False, False);
      if i mod 4 = 0 then UpdateProgress;
    end;
    for i := 1 to instances do
    begin
      FreeAndNil(normalizer[i]);
      if i mod 4 = 0 then UpdateProgress;
    end;
  end;
  
  WriteLn('Done.'#10);
  StartStopTest(False);
  
  MessageBox(self.WindowHandle, 'Test completed successfully.', PAnsiChar(self.Caption), MB_ICONINFORMATION);
end;

//-----------------------------------------------------------------------------
// TEST #2
//-----------------------------------------------------------------------------

procedure TDynamicAudioNormalizerTestApp.ButtonTest2Click(Sender: TObject);
const
  frameSize = 4096;
  channelCount = 2;
  MAX_SHORT: Double = 32767.0;
var
  samples: Array of TDoubleArray;
  buffer: array[0..((channelCount*frameSize)-1)] of SmallInt;
  inputFile, outputFile: THandle;
  normalizer: TDynamicAudioNormalizer;
  i, j, k, q: Cardinal;
  readSize, writeSize, bytesWritten: Cardinal;
begin
  {========================= SELECT FILES ========================}

  if not OpenDialog.Execute then
  begin
    Exit;
  end;

  if not SaveDialog.Execute then
  begin
    Exit;
  end;

  {============================ SETUP ============================}

  StartStopTest(True);
  
  SetLength(samples, channelCount, frameSize);

  WriteLn('Open I/O files...');
  inputFile  := CreateFile(PAnsiChar(OpenDialog.Files[0]),  GENERIC_READ, FILE_SHARE_READ, nil, OPEN_EXISTING, 0, 0);
  outputFile := CreateFile(PAnsiChar(SaveDialog.Files[0]), GENERIC_WRITE, 0, nil, CREATE_ALWAYS, 0, 0);

  if (inputFile = INVALID_HANDLE_VALUE) or (outputFile = INVALID_HANDLE_VALUE) then
  begin
    Raise Exception.Create('Failed to open input/output files!');
  end;

  WriteLn('Creating normalizer instance...');
  normalizer := TDynamicAudioNormalizer.Create(channelCount, 44100, 500, 31, 0.95, 10.0, 0.0, 0.0, True, False, False);
  WriteLn('Internal delay: ' + IntToStr(normalizer.GetInternalDelay));

  {=========================== PROCESS ===========================}

  WriteLn('Processing, please wait...');
  for q := 0 to 2147483647 do
  begin
    if not ReadFile(inputFile, buffer, frameSize * channelCount * SizeOf(SmallInt), readSize, nil) then
    begin
      Raise Exception.Create('Read operation has failed!');
    end;

    readSize := readSize div (channelCount * SizeOf(SmallInt));

    if readSize > 0 then
    begin
      k := 0;
      for i := 0 to (readSize-1) do
      begin
        for j := 0 to (channelCount-1) do
        begin
          samples[j][i] := buffer[k] / MAX_SHORT;
          k := k + 1;
        end;
      end;
    end;

    writeSize := normalizer.ProcessInplace(samples, readSize);
    //WriteLn('Result: readSize=' + IntToStr(readSize) + '; writeSize=' +  IntToStr(writeSize));

    if writeSize > 0 then
    begin
      k := 0;
      for i := 0 to (writeSize-1) do
      begin
        for j := 0 to (channelCount-1) do
        begin
          buffer[k] := SmallInt(Round(Max(-1.0,Min(1.0,samples[j][i])) * MAX_SHORT));
          k := k + 1;
        end;
      end;
      if not WriteFile(outputFile, buffer, writeSize * channelCount * SizeOf(SmallInt), bytesWritten, nil) then
      begin
        Raise Exception.Create('Write operation has failed!');
      end;
    end;

    if readSize < frameSize then
    begin
      Break; {End of File}
    end;

    if q mod 32 = 0 then
    begin
      UpdateProgress;
    end;
  end;

  {============================ FLUSH ============================}

  WriteLn('Flushing buffer, please wait...');
  for q := 0 to 2147483647 do
  begin
    writeSize := normalizer.FlushBuffer(samples);

    if writeSize > 0 then
    begin
      k := 0;
      for i := 0 to (writeSize-1) do
      begin
        for j := 0 to (channelCount-1) do
        begin
          buffer[k] := SmallInt(Round(Max(-1.0,Min(1.0,samples[j][i])) * MAX_SHORT));
          k := k + 1;
        end;
      end;
      if not WriteFile(outputFile, buffer, writeSize * channelCount * SizeOf(SmallInt), bytesWritten, nil) then
      begin
        Raise Exception.Create('Write operation has failed!');
      end;
    end;

    if writeSize < 1 then
    begin
      Break; {No more samples left}
    end;

    if q mod 32 = 0 then
    begin
      UpdateProgress;
    end;
  end;

  {============================ CLOSE ============================}

  WriteLn('Shutting down...');
  normalizer.Reset(); {Just to make sure that the Reset() function works}

  {============================ RESET ============================}

  FreeAndNil(normalizer);
  FlushFileBuffers(outputFile);

  CloseHandle(inputFile);
  CloseHandle(outputFile);

  WriteLn('Done.'#10);
  StartStopTest(False);

  MessageBox(self.WindowHandle, 'Test completed successfully.', PAnsiChar(self.Caption), MB_ICONINFORMATION);
end;

end.

File: DynamicAudioNormalizerNET/samples/DynamicAudioNormalizerExample.vb
REM //////////////////////////////////////////////////////////////////////////////////
REM // Dynamic Audio Normalizer - Microsoft.NET Wrapper
REM // Copyright (c) 2014-2018 LoRd_MuldeR <mulder2@gmx.de>. Some rights reserved.
REM //
REM // Permission is hereby granted, free of charge, to any person obtaining a copy
REM // of this software and associated documentation files (the "Software"), to deal
REM // in the Software without restriction, including without limitation the rights
REM // to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
REM // copies of the Software, and to permit persons to whom the Software is
REM // furnished to do so, subject to the following conditions:
REM //
REM // The above copyright notice and this permission notice shall be included in
REM // all copies or substantial portions of the Software.
REM //
REM // THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
REM // IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
REM // FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
REM // AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
REM // LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
REM // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
REM // THE SOFTWARE.
REM //
REM // http://opensource.org/licenses/MIT
REM //////////////////////////////////////////////////////////////////////////////////

Imports System.Reflection
Imports DynamicAudioNormalizer

Module DynamicAudioNormalizerExample

    REM --------------------------------------------------------------------------
    REM AudioFileReader Class
    REM --------------------------------------------------------------------------
    Class AudioFileReader
        Implements IDisposable
        Private Reader As IO.BinaryReader

        Public Sub New(FileName As String)
            Reader = New IO.BinaryReader(IO.File.OpenRead(FileName))
        End Sub

        Protected Sub Dispose() Implements IDisposable.Dispose
            If Reader IsNot Nothing Then
                Reader.Dispose()
            End If
        End Sub

        Public Function Read(SamplesOut(,) As Double) As Long
            Dim Channels As Integer = SamplesOut.GetLength(0)
            Dim MaxSamples As Integer = SamplesOut.GetLength(1)
            For I As Long = 0 To MaxSamples - 1
                For C As Long = 0 To Channels - 1
                    Try
                        SamplesOut(C, I) = CType(Reader.ReadInt16(), Double) / 32767.0
                    Catch E As IO.EndOfStreamException
                        Return I
                    End Try
                Next
            Next
            Return MaxSamples
        End Function
    End Class

    REM --------------------------------------------------------------------------
    REM AudioFileWriter Class
    REM --------------------------------------------------------------------------
    Class AudioFileWriter
        Implements IDisposable
        Private Writer As IO.BinaryWriter

        Public Sub New(FileName As String)
            Writer = New IO.BinaryWriter(IO.File.OpenWrite(FileName))
        End Sub

        Protected Sub Dispose() Implements IDisposable.Dispose
            If Writer IsNot Nothing Then
                Writer.Flush()
                Writer.Dispose()
            End If
        End Sub

        Public Sub Write(SamplesOut(,) As Double, SampleCount As Long)
            Dim Channels As Integer = SamplesOut.GetLength(0)
            For I As Long = 0 To SampleCount - 1
                For C As Long = 0 To Channels - 1
                    Writer.Write(CType(SamplesOut(C, I) * 32767.0, Short))
                Next
            Next
        End Sub
    End Class

    REM --------------------------------------------------------------------------
    REM Logging Callback
    REM --------------------------------------------------------------------------
    Private Sub LoggingHandler(Level As Int32, Message As String)
        Select Case Level
            Case 0
                Console.WriteLine("[DynAudNorm][DBG] " & Message)
            Case 1
                Console.WriteLine("[DynAudNorm][WRN] " & Message)
            Case 2
                Console.WriteLine("[DynAudNorm][ERR] " & Message)
            Case Else
                Console.WriteLine("[DynAudNorm][???] " & Message)
        End Select
    End Sub

    REM --------------------------------------------------------------------------
    REM Main Function
    REM --------------------------------------------------------------------------
    Sub Main()
        Dim Version As String
        Version = Assembly.GetExecutingAssembly().GetName().Version.ToString()
        Console.WriteLine("DynamicAudioNormalizer.NET VisualBasic Example [" & Version & "]" & vbCrLf)

        Try
            DynamicAudioNormalizerNET.setLogger(AddressOf LoggingHandler)

            Dim MajorVer As UInteger = 0, MinorVer As UInteger = 0, PatchVer As UInteger = 0
            DynamicAudioNormalizerNET.getVersionInfo(MajorVer, MinorVer, PatchVer)
            Console.WriteLine("Library Version: " & MajorVer & "." & Format(MinorVer, "##00") & "-" & PatchVer)

            Dim DateStr As String = "", TimeStr As String = "", CompilerStr As String = "", ArchStr As String = "", IsDebug As Boolean = False
            DynamicAudioNormalizerNET.getBuildInfo(DateStr, TimeStr, CompilerStr, ArchStr, IsDebug)
            Console.WriteLine("Library Build Date: " & DateStr)
            Console.WriteLine("Library Build Time: " & TimeStr)
            Console.WriteLine("Library Compiler: " & CompilerStr & " (" & ArchStr & ")")
            Console.WriteLine("Library Is Debug: " & IsDebug)

            Using Reader = New AudioFileReader("Input.pcm")
                Using Writer = New AudioFileWriter("Output.pcm")
                    Using Instance = New DynamicAudioNormalizerNET(2, 44100, 500, 31, 0.95, 10.0, 0.0, 0.0, True, False, False)
                        ProcessAudioFile(Instance, Reader, Writer)
                    End Using
                End Using
            End Using

            Console.WriteLine("Test completed." & vbCrLf)
        Catch Except As Exception
            Console.WriteLine(vbCrLf & "FAILURE !!!")
            Console.WriteLine(vbCrLf & Except.GetType().ToString() & ": " & Except.Message & vbCrLf)
        End Try
    End Sub

    REM --------------------------------------------------------------------------
    REM Processing Loop
    REM --------------------------------------------------------------------------
    Private Sub ProcessAudioFile(Instance As DynamicAudioNormalizerNET, Reader As AudioFileReader, Writer As AudioFileWriter)
        Dim Channels As UInteger = 0, FilterSize As UInteger = 0, FrameLen As UInteger = 0, SampleRate As UInteger = 0
        Instance.getConfiguration(Channels, FilterSize, FrameLen, SampleRate)
        Console.WriteLine("Channels: " & Channels)
        Console.WriteLine("FilterSize: " & FilterSize)
        Console.WriteLine("FrameLen: " & FrameLen)
        Console.WriteLine("SampleRate: " & SampleRate)

        Dim InternalDelay As Long = Instance.getInternalDelay()
        Console.WriteLine("InternalDelay: " & InternalDelay)

        REM Note: VisualBasic specifies the highest index value (i.e. Length - 1) rather than the Length, so the following creates a 2 x 4096 array!
        Dim AudioSamples = New Double(Channels - 1, 4095) {}

        Console.WriteLine("Processing audio samples, please wait...")
        While True
            Dim InputSamples = Reader.Read(AudioSamples)
            If InputSamples > 0 Then
                Dim OutputSamples = Instance.processInplace(AudioSamples, InputSamples)
                If OutputSamples > 0 Then
                    Writer.Write(AudioSamples, OutputSamples)
                End If
            End If
            If InputSamples < AudioSamples.GetLength(1) Then
                REM End of file reached!
                Exit While
            End If
        End While

        Console.WriteLine("Flushing buffer, please wait...")
        While True
            Dim OutputSamples = Instance.flushBuffer(AudioSamples)
            If OutputSamples > 0 Then
                Writer.Write(AudioSamples, OutputSamples)
                Continue While
            End If
            REM No more samples to flush!
            Exit While
        End While

        REM Only for testing, not usually required!
        Instance.reset()
    End Sub

End Module
